src/reconcile/rec_visibility.c

```diff
@@ -182,6 +182,134 @@ __rec_append_orig_value(
+        WT_ASSERT(session, hs_tw->stop_ts == delete_tombstone->start_ts);
+        WT_ASSERT(session, hs_tw->durable_stop_ts == delete_tombstone->durable_ts);
+    } else
+        WT_ASSERT(session, !WT_TIME_WINDOW_HAS_STOP(hs_tw));
+#endif
+
+    WT_ERR(r->hs_cursor->remove(r->hs_cursor));
+done:
+    if (delete_tombstone != NULL)
+        F_CLR(delete_tombstone, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+    F_CLR(delete_upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+
+err:
+    if (!hs_read_committed)
+        F_CLR(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+    return (ret);
+}
+
+/*
+ * __rec_find_and_delete_hs_record --
+ *     Find and delete the update left in the history store
+ */
+static int
+__rec_find_and_delete_hs_record(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins,
+  WT_ROW *rip, WT_UPDATE_SELECT *upd_select)
+{
+    WT_DECL_ITEM(key);
+    WT_DECL_RET;
+    WT_UPDATE *delete_tombstone, *delete_upd;
+    uint8_t *p;
+
+    delete_tombstone = NULL;
+
+    for (delete_upd = upd_select->tombstone != NULL ? upd_select->tombstone : upd_select->upd;
+         delete_upd != NULL; delete_upd = delete_upd->next) {
+        if (delete_upd->txnid == WT_TXN_ABORTED)
+            continue;
+
+        if (F_ISSET(delete_upd, WT_UPDATE_TO_DELETE_FROM_HS)) {
+            WT_ASSERT_ALWAYS(session,
+              F_ISSET(delete_upd, WT_UPDATE_HS | WT_UPDATE_RESTORED_FROM_HS),
+              "Attempting to remove an update from the history store in WiredTiger, but the "
+              "update was missing.");
+            if (delete_upd->type == WT_UPDATE_TOMBSTONE)
+                delete_tombstone = delete_upd;
+            else {
+                WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
+
+                switch (r->page->type) {
+                case WT_PAGE_COL_FIX:
+                case WT_PAGE_COL_VAR:
+                    p = key->mem;
+                    WT_ERR(__wt_vpack_uint(&p, 0, WT_INSERT_RECNO(ins)));
+                    key->size = WT_PTRDIFF(p, key->data);
+                    break;
+                case WT_PAGE_ROW_LEAF:
+                    if (ins == NULL) {
+                        WT_ERR(__wt_row_leaf_key(session, r->page, rip, key, false));
+                    } else {
+                        key->data = WT_INSERT_KEY(ins);
+                        key->size = WT_INSERT_KEY_SIZE(ins);
+                    }
+                    break;
+                default:
+                    WT_ERR(__wt_illegal_value(session, r->page->type));
+                }
+                WT_ERR(__rec_delete_hs_record(session, r, key, delete_upd, delete_tombstone));
+                if (r->page->type == WT_PAGE_COL_FIX)
+                    printf("Remove hs record %" PRIu64 "\n", WT_INSERT_RECNO(ins));
+                break;
+            }
+        }
+    }
+
+    /*
+     * If we see a restored tombstone from the history store, we must also see a restored update
```

Comment:
The `must` makes me wonder if we want instead `WT_ASSERT_ALWAYS`, what do you think?

Reply:
done